app/services/import/csv/activity_actual_refund_comment/row_service.rb

```diff
@@ -0,0 +1,135 @@
+class Import::Csv::ActivityActualRefundComment::RowService
+  attr_reader :errors
+
+  def initialize(report, user, row)
+    @report = report
+    @user = user
+    @activity = nil
+    @errors = {}
+    @row = Import::Csv::ActivityActualRefundComment::Row.new(row)
+  end
+
+  def import!
+    if @row.valid?
+      @activity = find_activity(@row.roda_identifier)
+
+      return false if @activity.nil?
+
+      return false unless authorise_activity(@activity)
+
+      return Import::Csv::ActivityActualRefundComment::SkippedRow.new(@row) if @row.empty?
+
+      create_record
+    else
+      @errors.update(@row.errors)
+      false
+    end
+  end
+
+  private def create_record
+    case type_of_row(@row)
+    when :actual
+      result = create_actual(@row, @activity)
+    when :refund
+      result = create_refund(@row, @activity)
+    when :comment
+      result = create_activity_comment(@row, @activity)
+    end
+
+    return false unless result
+
+    if result.failure?
+      result.object.errors.each do |error|
+        @errors[error.attribute] = [error.attribute, error.message]
+      end
+      false
+    else
+      result.object
+    end
+  end
+
+  private def find_activity(roda_identifier)
+    activity = Activity.find_by_roda_identifier(roda_identifier)
+
+    unless activity
+      @errors.update({"Activity RODA Identifier" => [
+        roda_identifier, I18n.t("import.csv.activity_actual_refund_comment.errors.activity_roda_identifier.not_found")
+      ]})
+    end
+
+    activity
+  end
+
+  private def authorise_activity(activity)
+    policy = ActivityPolicy.new(@user, activity)
+    if !policy.create?
+      @errors.update({"Activity RODA Identifier" => [
+        activity.roda_identifier,
+        I18n.t("import.csv.activity_actual_refund_comment.errors.activity_roda_identifier.not_authorised")
+      ]})
+      return false
+    elsif !reportable_activity?(activity)
+      @errors.update({"Activity RODA Identifier" => [
+        activity.roda_identifier,
+        I18n.t("import.csv.activity_actual_refund_comment.errors.activity_roda_identifier.not_reportable")
+      ]})
+      return false
+    end
+    true
+  end
+
+  private def reportable_activity?(activity)
+    activity.organisation == @report.organisation && activity.associated_fund == @report.fund
+  end
+
+  private def type_of_row(row)
+    return :actual if row.actual_value.positive? && row.refund_value.zero?
```

Comment:
This might be a preference, but this scenario leaves the situation where the actual_value is negative and the refund_value is zero uncovered.

If a user tries to input a negative value in the new actual form, they get an error message, "Value cannot be negative".

Here this would result in a row whose type cannot be identified, and is therefore ignored without an error message. My preference would be to show an error message if possible, rather than skipping it quietly.

Reply:
Nice one! 

Yep, I've missed that :( I would like the responsibility for that to belong to `Import::Csv::ActivityActualRefundComment::Row` so I will make a PR for that change to rebase this onto.

Reply:
@CristinaRO negative actual value validation PR #2303